### PR TITLE
Add nd-gaussians to Splatfacto

### PR DESCRIPTION
**Pull request by CardiacMangoes**
Adds support for rasterizing gaussians with extra dimensions

Adds parameters to gsplat:
- `use_nd` (bool)
- `nd_dim` (int)
- `nd_detach` (bool)

### CODE REVIEW

**Comment by kerrj** [comment]
I know that internally we've used the term "nd" a lot, but it's not very precise from an outside perspective. Can we rename things to "render_features" or "render_latents" or something like that? then instead of "nd_values" it could output something like "feature_map" which is more clear what's going on.

File: nerfstudio/models/splatfacto.py [hunk 1]
```diff
@@ -167,6 +167,12 @@ class SplatfactoModelConfig(ModelConfig):
     However, PLY exported with antialiased rasterize mode is not compatible with classic mode. Thus many web viewers that
     were implemented for classic mode can not render antialiased mode PLY properly without modifications.
     """
+    use_nd: bool = False
+    """Whether to rasterize nd values."""
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:171]: Expand on the docstrings a lot here, mention how each gaussians stores a feature vector of a certain dimension, and it gets rasterized as separate channels.

File: nerfstudio/models/splatfacto.py [hunk 2]
```diff
@@ -660,17 +707,25 @@ def get_outputs(self, camera: Cameras) -> Dict[str, Union[torch.Tensor, List]]:
...
         assert camera.shape[0] == 1, "Only one camera at a time"
 
+        out = {}
+
         # get the background color
+        dims = self.config.nd_dim if self.config.use_nd else 3
         if self.training:
             if self.config.background_color == "random":
                 background = torch.rand(3, device=self.device)
+                nd_background = torch.rand(dims, device=self.device)
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:717]: Using random background for training only really makes sense for RGB since the gradients help opacity inflate into uncovered regions of the scene. I think the behavior for feature vectors should be to always use a torch.zeros background, then normalize with alpha similar to depth. (or maybe this step can be left out as optional if people want to add it)

File: nerfstudio/models/splatfacto.py [hunk 3]
```diff
@@ -811,8 +874,24 @@ def get_outputs(self, camera: Cameras) -> Dict[str, Union[torch.Tensor, List]]:
                 background=torch.zeros(3, device=self.device),
             )[..., 0:1]  # type: ignore
             depth_im = torch.where(alpha > 0, depth_im / alpha, depth_im.detach().max())
+        if self.config.use_nd:
+            nd_values = self.get_nd_values()
+            nd_value = rasterize_gaussians(  # type: ignore
+                self.xys.detach() if self.config.nd_detach else self.xys,
+                depths,
```
Comment (kerrj) [inline nerfstudio/models/splatfacto.py:881]: depths should also have the same `.detach()` logic

**Comment by kerrj** [comment]
We should refactor this to integrate with #2946, it will simplify the logic.
